[Cloudflare] Use CEL input to add support for token authorization in Audit log data stream

taniumalloy opened the pull request:
… CEL and support auth tokens.

- Enhancement

## Proposed commit message

The previous version did not support token authorization and restricted users to key authorization only, which is a [legacy method](https://developers.cloudflare.com/fundamentals/api/get-started/keys/). Introducing support for token authorization enables admins to implement least privilege and better restrict the scope of the keys created. The legacy key authorization method issues global keys provisioned with the same access as the user/admin who creates them. 

To introduce support for tokens, the input type for the Audit data stream was converted from httpjson to CEL. I chose CEL due to its robustness and widespread use among many integrations, enabling me to implement best practices, including state tracking effectively. The Audit data stream manifest was updated to include a field for tokens, modeling the existing manifest for the logpull manifest. The CEL program tracks state, cursor, and redacts sensitive fields appropriately. Many options were included to enable administrator control over page size, polling intervals, and lookback periods in addition to the standard support for SSL, proxy, etc. 

The README.md file was updated to reflect the new capabilities in the Audit input.


<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [x] Validate the end-user can switch between key and the token authorization 
- [x] Validate the program handles errors effectively and displays informative error messages to the user
- [x] Validate the integration redacts sensitive auth fields
- [x] Validate the paging, interval, and lookback functionality works
- [x] Verify spelling and grammar

## How to test this PR locally

Use elastic-package to construct a local Dockerized cluster and test with Cloudflare API. Test with API key and token if possible, use request tracing to view changes to authorization headers, test edge boundaries, and test integration with built-in elastic-package testing commands. 

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- N/a

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

efd6 commented:
@taniumalloy Please sign the CLA so that we can take a look at this.

taniumalloy commented:
I just signed the agreement with my GitHub username. Please let me know if there are any issues.

efd6 reviewed (commented):
Has this been discussed? Is there an issue to track it?

This is a significant breaking change and I think we need to ensure that it is aligned before we go ahead with it.

efd6 commented:
/test

taniumalloy commented:
It has not been discussed, and there are no issues that I am aware of. I have not submitted an issue. I know this integration is supported by the community, so I was not sure how to approach getting these changes implemented other than submitting a PR. 

If you would like, we can close this PR, and I can proceed with submitting an issue. Please advise me of any other steps you would like me to take. 

Thank you

efd6 commented:
I'll raise this internally. Until then, we can pause this.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/25180)
* Commit: 9238c3dd572981f795ab381a1d2f1429e33d277d

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25180","number":25180,"commit":"9238c3dd572981f795ab381a1d2f1429e33d277d"}],"number":25180}
buildkite-pr-comment-->

efd6 commented:
/test

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/25183)
* Commit: 057f53a7cac4508ed74ebc203d545e31d488e3d7

### History
* :green_heart: [Build #25180](https://buildkite.com/elastic/integrations/builds/25180) succeeded 9238c3dd572981f795ab381a1d2f1429e33d277d

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25183","number":25183,"commit":"057f53a7cac4508ed74ebc203d545e31d488e3d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25180","number":25180,"commit":"9238c3dd572981f795ab381a1d2f1429e33d277d"}],"number":25183}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/security-service-integrations (Team:Security-Service Integrations)

efd6 commented:
@taniumalloy, thanks for your patience. We can't break people here, but we can add the new functionality using the CEL input. So please revert the deletion of the HTTP JSON input while retaining the new CEL input and mark the HTTP JSON input as deprecated and set the CEL input as `enabled: false`. Once that's done ping me and I'll review.

taniumalloy commented:
@efd6 Thank you for your assistance. I understand and think that is a good solution. Would you like me to move the PR to draft mode while I work on this?

efd6 commented:
Sure, move it to draft, and when you are ready, let me know.

taniumalloy commented:
@efd6 I am ready for you to take a look. Here is a summary of the changes I made:

- Updated cursor time tracking to use RFC3339Nano as some audit events, such as DNS changes, are logged with ms accuracy.
- Removed the +1s increment from the cursor time tracking. While keeping the same cursor timestamp for back-to-back batches with no new events will keep fetching the most recent event due to Cloudflare's inclusive "since" boundary, the pipeline assigns an _id value to each log based on a field set by Cloudflare that is unique for each event, so no duplicates are indexed. Based on what I see in other integrations, this seems to be common and acceptable. This also ensures the entire time range is covered to guarantee audit log delivery. 
- Added v1 tag to the parent httpjson input.
- Added v2 tag to the parent cel input.
- Disabled the cel input by default, ensuring current users will not be affected by the upgrade.
- Included a description for the deprecated HTTP JSON Audit logs input, informing users to migrate workload to the v2 input for Audit logs.
- Included a description for the v1 Cloudflare Logpull input, informing users to consider using Logpush. While Logull is a [legacy](https://developers.cloudflare.com/logs/logpull/) collection method, it is not yet deprecated by Cloudflare and is still supported. However, I feel this note is aligned well with the inherited v1 tag. 
- Made many modifications to the README.md to update broken links, improve phrasing, inform users of legacy methods, and organize the file to increase clarity and accuracy. 

I have conducted tests and see no functional issues. Upgrading the integration did not affect my established HTTP JSON Audit log input. I was able to move to the new v2 CEL Audit log input and change between Key and Token auth mechanisms without issue.

Let me know what you think. If you have any concerns or see any issues, please let me know. Thank you

efd6 commented on packages/cloudflare/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,181 @@
+                },
+                "account_id": state.account_id,
+                "per_page": state.per_page,
+                ?"auth_token": has(state.auth_token) ?
+                    optional.of(state.auth_token)
+                :
+                    optional.none(),
+                ?"auth_email": has(state.auth_email) ? 
+                    optional.of(state.auth_email)
+                :
+                    optional.none(),
+                ?"auth_key": has(state.auth_key) ? 
+                    optional.of(state.auth_key)
+                :
+                    optional.none(),
```
```suggestion
                ?"auth_token": state.?auth_token,
                ?"auth_email": state.?auth_email,
                ?"auth_key": state.?auth_key,
```

efd6 commented:
/test

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/25389)
* Commit: e5220b460a0043a45680a982887c2653a05914f2

### History
* :green_heart: [Build #25183](https://buildkite.com/elastic/integrations/builds/25183) succeeded 057f53a7cac4508ed74ebc203d545e31d488e3d7
* :green_heart: [Build #25180](https://buildkite.com/elastic/integrations/builds/25180) succeeded 9238c3dd572981f795ab381a1d2f1429e33d277d

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25389","number":25389,"commit":"e5220b460a0043a45680a982887c2653a05914f2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25183","number":25183,"commit":"057f53a7cac4508ed74ebc203d545e31d488e3d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25180","number":25180,"commit":"9238c3dd572981f795ab381a1d2f1429e33d277d"}],"number":25389}
buildkite-pr-comment-->

taniumalloy commented:
@efd6 Thanks for the detailed review and suggestions. I resolved all of your suggestions and fixes. 

Summary:
- Bumped up minimum Kibana version for v2.31.0 to 8.16.0 to get the newest features
- Updated the request and resource tracing module configuration in each stream config
- Implemented recommended CEL program changes
- Focused on Audit log input
- Changed parent HTTP JSON module naming to "Collect Cloudflare logs via API (Legacy Input)"

I kept the README.md changes. Let me know if you have issues with these changes.

The new parent CEL module is still named as "Collect Cloudflare logs via API (v2)", so let me know if you have any issues or suggestions. The (v2) made sense with the corresponding (v1) parent, but I am open to ideas now. I can leave it or remove the (v1) tag altogether.

efd6 commented:
/test

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/25595)
* Commit: 49e0431f2f184fb327b6fa003e08580f8d441487

### History
* :green_heart: [Build #25389](https://buildkite.com/elastic/integrations/builds/25389) succeeded e5220b460a0043a45680a982887c2653a05914f2
* :green_heart: [Build #25183](https://buildkite.com/elastic/integrations/builds/25183) succeeded 057f53a7cac4508ed74ebc203d545e31d488e3d7
* :green_heart: [Build #25180](https://buildkite.com/elastic/integrations/builds/25180) succeeded 9238c3dd572981f795ab381a1d2f1429e33d277d

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25595","number":25595,"commit":"49e0431f2f184fb327b6fa003e08580f8d441487"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25389","number":25389,"commit":"e5220b460a0043a45680a982887c2653a05914f2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25183","number":25183,"commit":"057f53a7cac4508ed74ebc203d545e31d488e3d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25180","number":25180,"commit":"9238c3dd572981f795ab381a1d2f1429e33d277d"}],"number":25595}
buildkite-pr-comment-->

efd6 commented on packages/cloudflare/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,161 @@
+                            "last_timestamp": state.cursor.last_timestamp,                           
+                        }
+                    :
+                    
```
```suggestion
```

efd6 commented on packages/cloudflare/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,161 @@
+                ?"auth_email": state.?auth_email,
+                ?"auth_key": state.?auth_key,
+            })
+
```
```suggestion
```

efd6 commented on packages/cloudflare/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,161 @@
+                    "error": {
+                        "code": string(resp.StatusCode),
+                        "id": string(resp.Status),
+                        "message": "GET " + state.url.trim_right("/") + "/client/v4/accounts/" + state.account_id + "/audit_logs:"+(
```
```suggestion
                        "message": "GET " + state.url.trim_right("/") + "/client/v4/accounts/" + state.account_id + "/audit_logs: "+(
```

efd6 reviewed (commented):
Nits only. Looking good. Thanks

efd6 commented:
/test

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/25695)
* Commit: 69000be83ecc7e3761ee77e43106d4846bf69084

### History
* :green_heart: [Build #25595](https://buildkite.com/elastic/integrations/builds/25595) succeeded 49e0431f2f184fb327b6fa003e08580f8d441487
* :green_heart: [Build #25389](https://buildkite.com/elastic/integrations/builds/25389) succeeded e5220b460a0043a45680a982887c2653a05914f2
* :green_heart: [Build #25183](https://buildkite.com/elastic/integrations/builds/25183) succeeded 057f53a7cac4508ed74ebc203d545e31d488e3d7
* :green_heart: [Build #25180](https://buildkite.com/elastic/integrations/builds/25180) succeeded 9238c3dd572981f795ab381a1d2f1429e33d277d

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25695","number":25695,"commit":"69000be83ecc7e3761ee77e43106d4846bf69084"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25595","number":25595,"commit":"49e0431f2f184fb327b6fa003e08580f8d441487"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25389","number":25389,"commit":"e5220b460a0043a45680a982887c2653a05914f2"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25183","number":25183,"commit":"057f53a7cac4508ed74ebc203d545e31d488e3d7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/25180","number":25180,"commit":"9238c3dd572981f795ab381a1d2f1429e33d277d"}],"number":25695}
buildkite-pr-comment-->

taniumalloy commented:
Thank you for all your help. Please let me know if you need anything else.

efd6 reviewed (approved):
Thanks
